### PR TITLE
Implement checkpoint fraud proofs

### PR DESCRIPTION
**Pull request by yorhodes**
### Description

Implements 4 categories of checkpoint fraud proofs for use in future validator slashing protocol:

1. **premature**: if a checkpoint index is greater than the corresponding mailbox count, it is fraudulent
2. **non local**: if a checkpoint origin does not match the checkpoint's mailbox domain, it is fraudulent
3. **message id**: if a checkpoint message ID differs from the actual message ID (verified by merkle proof) at the checkpoint index, it is fraudulent
4. **root**: if a checkpoint root differs from the actual root (verified by merkle proof + root reconstruction) at the checkpoint index, it is fraudulent

Notably this is implemented independently from signature verification to allow for multiple checkpoint signing schemes to reuse the same checkpoint logic.

### Related issues

- Fixes #[issue number here]

### Backward compatibility

Yes

### Testing

Unit Tests

### CODE REVIEW

**Comment by nambrot** [comment]
@mattiecnvr i assume you did not intend on closing this one

File: solidity/test/CheckpointFraudProofs.t.sol [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        );
+
+        bytes32 actualMessageId = checkpoint.messageId;
+        checkpoint.messageId = ~checkpoint.messageId;
+        assertTrue(
+            cfp.isFraudulentMessageId(checkpoint, proof, actualMessageId)
+        );
+    }
+
+    function testIsFraudulentRoot() public {
```
Comment (asaj) [inline solidity/test/CheckpointFraudProofs.t.sol:71]: Let i_i = checkpoint index and i_j = cached root index, r_i = correct checkpoint root, r!_i = fraudulent checkpoint root, and r_j = cached root

I would like to see a test where i_i != i_j, and where we test that we are able to properly recover r_i from r_j + a proof of m_i

**Review by asaj — APPROVED** [review]
Very nice work, mostly nits

**Comment by avious00** [comment]
@yorhodes @nambrot where is this story at?

**Comment by avious00** [comment]
moving this to the backlog, to be revisited when staking is implemented

**Comment by yorhodes** [comment]
https://github.com/dmfxyz/murky

**Pull request closed without merging**